### PR TITLE
Fixes #3929 - Deadlock between new HTTP2Connection() and Server.stop().

### LINKED ISSUE #3929: Deadlock between new HTTP2Connection() and Server.stop()
`Server.stop()` holds `AbstractLifeCycle._lock` while it joins a network thread:

```
  java.lang.Thread.State: WAITING
	 blocks qtp2102368942-24@2434
	  at java.lang.Object.wait(Object.java:-1)
	  at java.lang.Thread.join(Thread.java:1313)
	  at org.eclipse.jetty.util.thread.QueuedThreadPool.doStop(QueuedThreadPool.java:179)
	  at org.eclipse.jetty.util.component.AbstractLifeCycle.stop(AbstractLifeCycle.java:89)
	  - locked <0xe86> (a java.lang.Object)
	  at org.eclipse.jetty.util.component.ContainerLifeCycle.stop(ContainerLifeCycle.java:178)
	  at org.eclipse.jetty.util.component.ContainerLifeCycle.doStop(ContainerLifeCycle.java:199)
	  at org.eclipse.jetty.server.handler.AbstractHandler.doStop(AbstractHandler.java:124)
	  at org.eclipse.jetty.server.Server.doStop(Server.java:464)
	  at org.eclipse.jetty.util.component.AbstractLifeCycle.stop(AbstractLifeCycle.java:89)
	  - locked <0xe87> (a java.lang.Object)
```

But that network thread will never complete because it’s attempting to acquire `AbstractLifeCycle._lock`:

```
"qtp2102368942-24@2434" prio=5 tid=0x18 nid=NA waiting for monitor entry
  java.lang.Thread.State: BLOCKED
	 waiting for Time-limited test@1899 to release lock on <0xe86> (a java.lang.Object)
	  at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:65)
	  at org.eclipse.jetty.util.component.ContainerLifeCycle.start(ContainerLifeCycle.java:167)
	  at org.eclipse.jetty.util.component.ContainerLifeCycle.doStart(ContainerLifeCycle.java:119)
	  at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:68)
	  - locked <0xe88> (a java.lang.Object)
	  at org.eclipse.jetty.util.component.LifeCycle.start(LifeCycle.java:139)
	  at org.eclipse.jetty.http2.HTTP2Connection.<init>(HTTP2Connection.java:70)
	  at org.eclipse.jetty.http2.server.HTTP2ServerConnection.<init>(HTTP2ServerConnection.java:100)
	  at org.eclipse.jetty.http2.server.AbstractHTTP2ServerConnectionFactory.newConnection(AbstractHTTP2ServerConnectionFactory.java:243)
	  at org.eclipse.jetty.server.NegotiatingServerConnection.onFillable(NegotiatingServerConnection.java:130)
	  at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:305)
	  at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:103)
	  at org.eclipse.jetty.io.ssl.SslConnection$DecryptedEndPoint.onFillable(SslConnection.java:427)
	  at org.eclipse.jetty.io.ssl.SslConnection.onFillable(SslConnection.java:321)
	  at org.eclipse.jetty.io.ssl.SslConnection$2.succeeded(SslConnection.java:159)
	  at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:103)
	  at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:117)
	  at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:698)
	  at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:804)
	  at java.lang.Thread.run(Thread.java:834)
```

So we get a deadlock.

Here’s [thread_dump.txt](https://github.com/eclipse/jetty.project/files/3458156/thread_dump.txt), captured in the deadlock.

I was able to reproduce this with [JettyFailsToStopTest.java](https://github.com/eclipse/jetty.project/files/3458159/JettyFailsToStopTest.java.txt). It needs OkHttp and Okio to create an HTTP/2 connection. I observed the deadlock 1 attempt out of every 5 on my MacBook Pro.

### PR DESCRIPTION
**Pull request by sbordet**
#3929.

Previously beans were started if they were not running, which
included them being in a STOPPING state.
Now beans are only started if they are in STOPPED state.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

### CODE REVIEW

File: jetty-util/src/main/java/org/eclipse/jetty/util/component/ContainerLifeCycle.java [hunk 1]
```diff
@@ -106,18 +106,20 @@ protected void doStart() throws Exception
...
                         case AUTO:
-                            if (l.isRunning())
-                                unmanage(b);
-                            else
+                            if (l.isStopped())
                             {
                                 manage(b);
                                 start(l);
                             }
+                            else
```
Comment (joakime) [inline jetty-util/src/main/java/org/eclipse/jetty/util/component/ContainerLifeCycle.java]: Wouldn't this change make state FAILED trigger `unmanage(b)` now?
(Something that didn't happen before)

**Comment by joakime** [comment]
The proposed change to `doStart()`

Managed Mode | State | Action Before | Action Now
-- | -- | -- | --
MANAGED | STARTING | - | -
  | STARTED | - | -
  | STOPPING | start(l) | -
  | STOPPED | start(l) | start(l)
  | FAILED | start(l) | -
AUTO | STARTING | unmanage(b) | unmanage(b)
  | STARTED | unmanage(b) | unmanage(b)
  | STOPPING | manage(b);start(l) | unmanage(b)
  | STOPPED | manage(b);start(l) | manage(b);start(l)
  | FAILED | manage(b);start(l) | unmanage(b)

So, STOPPING and FAILED no longer attempt to `start(l)`, and in the case of AUTO now call `unmanage(b)` as well.

File: jetty-util/src/main/java/org/eclipse/jetty/util/component/ContainerLifeCycle.java [hunk 2]
```diff
@@ -106,18 +106,20 @@ protected void doStart() throws Exception
                     switch (b._managed)
                     {
                         case MANAGED:
-                            if (!l.isRunning())
+                            if (l.isStopped())
```
Comment (gregw) [inline jetty-util/src/main/java/org/eclipse/jetty/util/component/ContainerLifeCycle.java]: Looking at the state table from @joakime , I think we also need to start here if the bean is `FAILED`.  It is only if it is in the transit state of `STOPPING` that we should no longer start it.

File: jetty-util/src/main/java/org/eclipse/jetty/util/component/ContainerLifeCycle.java (hunk 1 above)
Comment (gregw) [reply jetty-util/src/main/java/org/eclipse/jetty/util/component/ContainerLifeCycle.java]: I agree that `FAILED` should be treated like `STOPPED`

**Comment by sbordet** [comment]
@gregw I don't understand why you want to treat FAILED like STOPPED.
If it's in FAILED state, it's either because it could not start or could not stop. 
Either case the component is now in a bogus state and I don't see how it can be reliably re-started.
I maintain that the current changes are correct.
I have added a test case.

**Comment by joakime** [comment]
Had a conversation with @sbordet about this change.

He's managed to convince me that the FAILED state behavior change on `Container.doStart()` is sane and valid still.

Some notes taken during the conversation.

Here's what the original intent of the PR was targeting as a fix.

```
  Alpha A1 = new Alpha();
  A1.start() => started
  Beta B1 = new Beta(); // (Container)

  B1.addBean(A1)
  B1.start()

   // Both B1 and A1 are now "started"

  Beta B2 = new Beta()
  A1.stop()

  during A1.doStop()
   {
     B2.addBean(this /* A1 */)
     B2.start()
   }
```

After going through a few what-if scenarios, the FAILED state still holds to its original meaning and behaviors (as this PR only changes Container.doStart() )

Example scenario:

```
  Beta B1 = new Beta(); // is a container - stopped
  Alpha A1 = new Alpha();
  Alpha A2 = new AlphaAlt();
  Alpha A3 = new AlphaFoo();
 
  B1.addBean(A1);
  B1.addBean(A2);
  B1.addBean(A3);

  -- STATE --
  B1 - stopped
    A1 - stopped - AUTO
    A2 - stopped - AUTO
    A3 - stopped - AUTO

  B1.start();

  A2.doStart() - fails

  -- STATE --
  B1 - failed
    A1 - MANAGED - started
    A2 - MANAGED - failed
    A3 - AUTO - stopped

  B1.stop();
    B1.doStop()
      A1 - managed - A1.stop() => stopped
      A2 - managed - A2.stop() => stopped
      A3 - auto - not A3.stop() - stays in stopped

  -- STATE --
  B1 - stopped
    A1 - stopped - MANAGED
    A2 - stopped - MANAGED
    A3 - stopped - AUTO

  B1.start()
    B1.doStart()
       A1 - managed - A1.start()
       A2 - managed - A2.start()
       A3 - auto - manage(b) ; A3.start()
```

**Comment by sbordet** [comment]
My point being that the FAILED state cannot be treated like the STOPPED state.

In FAILED state it is obvious that the bean lifecycle is managed externally (otherwise it could not have gone into FAILED).

In STOPPED state it's unclear because it could be a pristine lifecycle that was never started, or a lifecycle that was started _and_ stopped externally.
Handling this difference is an orthogonal issue.

But there is a difference between these 2 states, so they cannot be treated the same.

**Comment by gregw** [comment]
I disagree.

In a dynamic scenario (let's say you are changing things via JMX), how does a FAILED component ever leave that state.  Previously we were able to start a failed component and retry starting it.  Now once it is FAILED, it will behave differently.

If A is managing B and B is failed, then starting A should start B.   If fact the only time that starting A should not start B is if B is in one of the "ing" states or already started... and all of those are close to ISE if A really is managing B!

So I'm still -1 on this.

**Comment by gregw** [comment]
After hanging out with @joakime I found a good example that justifies my -1.  If we have the state:
```
 B1 - failed
    A1 - MANAGED - started
    A2 - MANAGED - failed
    A3 - AUTO - stopped
```
and we call start on B1, this PR as currently proposed would result in:
``` 
B1 - started
    A1 - MANAGED - started
    A2 - MANAGED - failed
    A3 - MANAGED - started
```
Which is nonsensical! B1 is started even though it has a failed component.

Either we retain the current behaviour for managed failed components OR we propose that no component can be started from failed (so B1.start() would ISE).  But that later proposal, while plausible, is entirely unrelated to the current issue and doesn't belong in this PR.... plus would need very careful consideration for inadvertent consequences.

If we revert the managed/failed behaviour to calling start, the above scenarios results in
``` 
B1 - started
    A1 - MANAGED - started
    A2 - MANAGED - started
    A3 - MANAGED - started
```
which is entirely sane.

**Pull request merged**